components/cop_codegen/src/rpn_function.rs

```diff
@@ -1,49 +1,321 @@
+                    ));
+                }
+            }
+        }
+        Ok(Self { is_varg, captures })
+    }
+}
+
+mod kw {
+    syn::custom_keyword!(Option);
+}
+
+/// Parses an evaluable type like `Option<T>`.
+struct RpnFnEvaluableType {
+    eval_type: TypePath,
+}
+
+impl parse::Parse for RpnFnEvaluableType {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        input.parse::<self::kw::Option>()?;
+        input.parse::<Token![<]>()?;
+        let eval_type = input.parse::<TypePath>()?;
+        input.parse::<Token![>]>()?;
+        Ok(Self { eval_type })
+    }
+}
+
+/// Parses a function signature parameter like `val: &Option<T>`.
+struct RpnFnSignatureParam {
+    _pat: Pat,
+    eval_type: TypePath,
+}
+
+impl parse::Parse for RpnFnSignatureParam {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        let pat = input.parse::<Pat>()?;
+        input.parse::<Token![:]>()?;
+        input.parse::<Token![&]>()?;
+        let et = input.parse::<RpnFnEvaluableType>()?;
+        Ok(Self {
+            _pat: pat,
+            eval_type: et.eval_type,
+        })
+    }
+}
+
+/// Parses a function signature parameter like `val: &[&Option<T>]`.
+struct VargsRpnFnSignatureParam {
+    _pat: Pat,
+    eval_type: TypePath,
+}
+
+impl parse::Parse for VargsRpnFnSignatureParam {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        let pat = input.parse::<Pat>()?;
+        input.parse::<Token![:]>()?;
+        input.parse::<Token![&]>()?;
+        let slice_inner;
+        bracketed!(slice_inner in input);
+        slice_inner.parse::<Token![&]>()?;
+        let et = slice_inner.parse::<RpnFnEvaluableType>()?;
+        Ok(Self {
+            _pat: pat,
+            eval_type: et.eval_type,
+        })
+    }
+}
+
+/// Parses a function signature return type like `Result<Option<T>>`.
+struct RpnFnSignatureReturnType {
+    eval_type: TypePath,
+}
+
+impl parse::Parse for RpnFnSignatureReturnType {
+    fn parse(input: parse::ParseStream<'_>) -> Result<Self> {
+        input.parse::<Token![->]>()?;
+        let tp = input.parse::<Type>()?;
+        if let Type::Path(TypePath {
+            path: Path { segments, .. },
+            ..
+        }) = &tp
+        {
+            let result_type = (*segments.last().unwrap().value()).clone();
+            if let PathArguments::AngleBracketed(AngleBracketedGenericArguments { args, .. }) =
+                result_type.arguments
+            {
+                let et = parse2::<RpnFnEvaluableType>(args.into_token_stream())?;
+                Ok(Self {
+                    eval_type: et.eval_type,
+                })
+            } else {
+                return Err(Error::new_spanned(
+                    tp,
+                    "expect angle bracketed path arguments",
+                ));
+            }
+        } else {
+            return Err(Error::new_spanned(tp, "expect path"));
+        }
+    }
+}
+
+pub struct RpnFn;
```

Comment:
Will it contain some fields in the future? Otherwise, why not make `transform` a free function?